src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs

```diff
@@ -1,442 +1,199 @@
+        /// <param name="innerRule">Inner rule. </param>
+        /// <returns></returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic")]
+        public IBindingProvider AddTypeFilter(Func<Type, bool> predicate, IBindingProvider innerRule)
         {
-            if (elementType == typeof(TMessage))
-            {
-                return new SimpleTriggerArgumentBinding<TMessage, TTriggerValue>(bindingStrategy, converterManager);
-            }
-            if (elementType == typeof(string))
-            {
-                return new StringTriggerArgumentBinding<TMessage, TTriggerValue>(bindingStrategy, converterManager);
-            }
-            else
-            {
-                // Default, assume a Poco
-                return new PocoTriggerArgumentBinding<TMessage, TTriggerValue>(bindingStrategy, converterManager, elementType);
-            }
+            return new GenericBinderFilter(predicate, innerRule);
+        }
+        
+        /// <summary>
+        /// Create a rule that returns an IValueBinder from a resolved attribute. IValueBinder will let you have an OnCompleted hook that 
+        /// is invoked after the user function completes. 
+        /// </summary>
+        /// <typeparam name="TAttribute"></typeparam>
+        /// <param name="builder"></param>
+        /// <returns></returns>
+        public IBindingProvider BindToGenericItem<TAttribute>(Func<TAttribute, Type, Task<IValueBinder>> builder)
+            where TAttribute : Attribute
+        {
+            return new GenericItemBindingProvider<TAttribute>(this._nameResolver, builder);
         }
 
         /// <summary>
-        /// Creates a binding that binds to an <see cref="IAsyncCollector{T}"/> with the specified argument type. Allows for a binding
-        /// to be generated for any POCO type.
+        /// Create a rule for binding a parameter to an <see cref="IAsyncCollector{TMEssage}"/>. 
+        /// Use the <see cref="IConverterManager"/> to convert form the user's parameter type to the TMessage type. 
         /// </summary>
-        /// <param name="parameter">The ParameterInfo for the binding</param>
-        /// <param name="collectorGenericType">The generic type that must implement <see cref="IAsyncCollector{T}"/> and have a public constructor with at most one parameter.</param>
-        /// <param name="collectorGenericArgumentType">The generic argument type for the <see cref="IAsyncCollector{T}"/></param>
-        /// <param name="converterManager">A converter manager to pass along.</param>        
-        /// <param name="invokeStringBinder">A <see cref="Func{T, TContext}"/> that returns the TContext to be used by the <see cref="IAsyncCollector{T}"/></param>
+        /// <typeparam name="TAttribute">type of binding attribute</typeparam>
+        /// <typeparam name="TMessage">'core type' for the IAsyncCollector.</typeparam>
```

Comment:
Same here and throughout - recommend finding a better term than "message" for this. Perhaps TElement or TItem or something general like that?